### PR TITLE
representation learner inside SAC

### PR DESCRIPTION
**Pull request by emailweixu**
With this, a target model of the representation can be maintained and be used to calculate the target critics. This is the strategy used by many other algorithms involving TD learning, (e.g. CURL, MuZero, etc).

### CODE REVIEW

**Review by Haichao-Zhang — DISMISSED** [review]
Thanks for the change! The proposed change properly handles the target version of the representation part, which is good in the presence of a representation module. 

I'm thinking of whether it is possible to achieve this goal within the ``Agent`` and/or ``ReprLearningAlgorithm`` class instead of ``SacAlgorithm``, since in that way, we can still reuse our previous modular design in terms of the functionality of each module in Agent. This could also make it easier to apply this Repr learning algorithm to other RL algorithm beyond SAC.

File: alf/algorithms/sac_algorithm.py [hunk 1]
```diff
@@ -270,6 +281,23 @@ def __init__(self,
...
+                observation_spec=observation_spec,
+                action_spec=action_spec,
+                reward_spec=reward_spec,
+                config=config)
+            target_repr_alg = repr_alg_ctor(
+                observation_spec=observation_spec,
+                action_spec=action_spec,
+                reward_spec=reward_spec,
+                config=config)
+            observation_spec = repr_alg.output_spec
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:298]: A repr alg doesn't necessarily contain a property called "output_spec"? Use assertion first will be better.

File: alf/algorithms/sac_algorithm.py [hunk 2]
```diff
@@ -773,7 +847,8 @@ def _critic_train_step(self, inputs: TimeStep, state: SacCriticState,
             target_critics = torch.sum(
                 discrete_act_dist.probs * target_critics, dim=-1)
 
-        target_critic = target_critics.reshape(inputs.reward.shape)
+        target_critic = target_critics.reshape(target_critics.shape[0],
+                                               *self._reward_spec.shape)
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:853]: Why is this change necessary?

File: alf/algorithms/sac_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/sac_algorithm.py:298]: added

File: alf/algorithms/sac_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/sac_algorithm.py:853]: This is because inputs.reward is no longer available

**Review by hnyu — APPROVED** [review]
